larray_editor/arrayadapter.py

```diff
@@ -3,188 +3,467 @@
-            if np.isscalar(filtered_bg_value):
-                filtered_bg_value = la.aslarray(filtered_bg_value)
-            return filtered_bg_value.data.reshape(shape_2D)
-        else:
-            return None
 
-    # XXX: or create two methods?:
-    # - set_data (which reset the current filter)
-    # - update_data (which sets new data but keeps current filter unchanged)
-    def set_data(self, data, bg_value=None):
-        assert isinstance(data, la.LArray)
+class AbstractAdapter(object):
+    def __init__(self, data, changes, bg_value):
+        self.data = data
+        self.bg_value = bg_value
+        self.changes = changes
         self.current_filter = {}
-        self.changes = {}
-        self.la_data = la.aslarray(data)
-        self.bg_value = la.aslarray(bg_value) if bg_value is not None else None
         self.update_filtered_data()
-        self.data_model.reset_minmax()
-        self.data_model.reset()
 
-    def update_filtered_data(self):
-        assert isinstance(self.la_data, la.LArray)
-        self.filtered_data = self.la_data[self.current_filter]
-
-        if np.isscalar(self.filtered_data):
-            self.filtered_data = la.aslarray(self.filtered_data)
-
-        axes = self.get_axes()
-        xlabels = self.get_xlabels()
-        ylabels = self.get_ylabels()
-        data_2D = self.get_2D_data()
-        changes_2D = self.get_changes_2D()
-        bg_value_2D = self.get_bg_value_2D(data_2D.shape)
-
-        self.axes_model.set_data(axes)
-        self.xlabels_model.set_data(xlabels)
-        self.ylabels_model.set_data(ylabels)
-        # using the protected version of the method to avoid calling reset() several times
-        self.data_model._set_data(data_2D)
-        self.data_model._set_changes(changes_2D)
-        self.data_model._set_bg_value(bg_value_2D)
-
-    def get_data(self):
-        return self.la_data
+    # ===================== #
+    #      PROPERTIES       #
+    # ===================== #
+
+    @property
+    def data(self):
+        return self._original_data
+
+    @data.setter
+    def data(self, original_data):
+        assert original_data is not None, "{} does not accept None as input data".format(self.__class__)
+        self._original_data = self.prepare_data(original_data)
+
+    @property
+    def bg_value(self):
+        return self._bg_value
+
+    @bg_value.setter
+    def bg_value(self, bg_value):
+        self._bg_value = self.prepare_bg_value(bg_value)
+
+    @property
+    def changes(self):
+        return self._changes
+
+    @changes.setter
+    def changes(self, changes):
+        if changes is None:
+            self._changes = {}
+        else:
+            assert isinstance(changes, dict), "{} only accept None or a dict as input changes".format(self.__class__)
+            self._changes = changes
 
     @property
     def ndim(self):
-        return self.filtered_data.ndim
+        return self.get_ndim(self._original_data)
```

Comment:
It would be simpler to make the ndim, size and dtype attributes themself the public API that the adapter must provide. Whether it provides it via a property or an instance attribute is up to the adapter writer.